### PR TITLE
Add isTrivial() and isTriviallyCopyable() AST matchers

### PR DESCRIPTION
**Pull request by higher-performance**
There are currently no AST matchers for the equivalent of `std::is_trivial`, `std::is_trivially_default_constructible`, `std::is_trivially_copyable`, and the like.

This pull request adds matchers to check if records and their member functions are trivial.

It does _not_ attempt to handle other trivial entities (such as pointers).

### CODE REVIEW

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 1]
```diff
@@ -5444,6 +5444,43 @@ AST_MATCHER(FunctionDecl, isDefaulted) {
   return Node.isDefaulted();
 }
 
+/// Matches trivial methods and types.
```
Comment (ilya-biryukov) [inline clang/include/clang/ASTMatchers/ASTMatchers.h:5447]: I do not think is actually matches `types`, but rather `classes` or `CXXRecordDecl`.
And the question is: should it?

I believe we could extend this matcher to also work on types and type locs.
Do you see any downsides to that? (same for the other matcher)

**Comment by ilya-biryukov** [comment]
I remember back in the day that adding new matchers was a problem because it grew the size of the files beyond some supported compiler's limit.

Could @AaronBallman or someone else knowledgable about this(cc @kadircet) take a look and confirm adding matchers is fine now?

**Review by ilya-biryukov — CHANGES_REQUESTED** [review]

**Comment by AaronBallman** [comment]
> I remember back in the day that adding new matchers was a problem because it grew the size of the files beyond some supported compiler's limit.
> 
> Could @AaronBallman or someone else knowledgable about this(cc @kadircet) take a look and confirm adding matchers is fine now?

We generally only add matchers when there's a need for them to reside in-tree (e.g., a few clang-tidy checks will use them); otherwise we recommend that folks define their own custom matchers. The reason is mostly because compile time overhead for compiling ASTMatchers.h is... huge... due to the template instantiations, so we usually want there to be a need before adding any.

Is there a planned need here?

**Comment by AaronBallman** [comment]
Thanks for the details! I'd say let's hold off on adding these matchers; if you have a check that needs them, then you could look through clang-tidy to see if other checks have also added a local matcher for the same thing, and if there's two or more matchers needing it, then we'll add it as a global matcher.

As for the check itself, that looks like an interesting one!

**Comment by philnik777** [comment]
While the libc++ checks are in the same tree as clang-tidy itself we (libc++) don't actually consider them in-tree, since they are compiled as a plugin against a clang-tidy that's already installed on the system.

**Pull request closed without merging**